File: backend/utils/jwt_helper.py
"""
Simple HS256 JWT Implementation to avoid PyJWT dependency in Cloudflare Workers.
"""

import base64
import hmac
import hashlib
import json
import time

def base64url_encode(payload):
    if isinstance(payload, str):
        payload = payload.encode('utf-8')
    elif isinstance(payload, dict):
        payload = json.dumps(payload).encode('utf-8')
    return base64.urlsafe_b64encode(payload).replace(b'=', b'').decode('utf-8')

def base64url_decode(payload):
    if isinstance(payload, str):
        payload = payload.encode('utf-8')
    rem = len(payload) % 4
    if rem > 0:
        payload += b'=' * (4 - rem)
    return base64.urlsafe_b64decode(payload)

class JWTError(Exception):
    pass

class PyJWTError(JWTError):
    pass

class ExpiredSignatureError(PyJWTError):
    pass

class InvalidTokenError(PyJWTError):
    pass


def encode(payload, secret, algorithm="HS256"):
    if algorithm != "HS256":
        raise ValueError("Only HS256 is supported")
    header = {"alg": "HS256", "typ": "JWT"}
    
    segments = []
    segments.append(base64url_encode(header))
    segments.append(base64url_encode(payload))
    
    signing_input = ".".join(segments).encode('utf-8')
    key = secret.encode('utf-8')
    signature = hmac.new(key, signing_input, hashlib.sha256).digest()
    
    segments.append(base64url_encode(signature))
    return ".".join(segments)

def decode(jwt_str, secret, algorithms=["HS256"], options=None):
    if "HS256" not in algorithms:
        raise ValueError("Only HS256 is supported")
    
    options = options or {}
    verify_exp = options.get("verify_exp", True)
    
    try:
        if isinstance(jwt_str, bytes):
            jwt_str = jwt_str.decode('utf-8')
        parts = jwt_str.split('.')
        if len(parts) != 3:
            raise InvalidTokenError("Invalid token segments")
        
        header_segment, payload_segment, crypto_segment = parts
        
        # Verify signature
        signing_input = (header_segment + "." + payload_segment).encode('utf-8')
        key = secret.encode('utf-8')
        signature = base64url_decode(crypto_segment)
        
        expected_signature = hmac.new(key, signing_input, hashlib.sha256).digest()
        
        if not hmac.compare_digest(signature, expected_signature):
            raise InvalidTokenError("Signature verification failed")
            
        payload = json.loads(base64url_decode(payload_segment).decode('utf-8'))
        
        if verify_exp and 'exp' in payload:
            if time.time() > payload['exp']:
                raise ExpiredSignatureError("Token expired")
                
        return payload
    except Exception as e:
        if isinstance(e, (ExpiredSignatureError, InvalidTokenError)):
            raise e
        raise InvalidTokenError(str(e))

File: backend/utils/jwt_helper.js
function str2ab(str) {
  return new TextEncoder().encode(str);
}

function bufferToBase64Url(buf) {
  const binstr = Array.from(new Uint8Array(buf), ch => String.fromCharCode(ch)).join('');
  const b64 = btoa(binstr);
  return b64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '');
}

function base64UrlDecode(str) {
  str = str.replace(/-/g, '+').replace(/_/g, '/');
  while (str.length % 4) str += '=';
  const binstr = atob(str);
  const buf = new Uint8Array(binstr.length);
  for (let i = 0; i < binstr.length; i++) {
    buf[i] = binstr.charCodeAt(i);
  }
  return buf;
}

export async function encode(payload, secret) {
  const header = { alg: "HS256", typ: "JWT" };
  const headerB64 = bufferToBase64Url(str2ab(JSON.stringify(header)));
  const payloadB64 = bufferToBase64Url(str2ab(JSON.stringify(payload)));
  const signingInput = `${headerB64}.${payloadB64}`;
  
  const key = await crypto.subtle.importKey(
    "raw",
    str2ab(secret),
    { name: "HMAC", hash: { name: "SHA-256" } },
    false,
    ["sign"]
  );
  
  const signature = await crypto.subtle.sign(
    "HMAC",
    key,
    str2ab(signingInput)
  );
  
  return `${signingInput}.${bufferToBase64Url(signature)}`;
}

export async function decode(jwtStr, secret, options = {}) {
  const verifyExp = options.verifyExp !== false;
  const parts = jwtStr.split('.');
  if (parts.length !== 3) {
    throw new Error("Invalid token segments");
  }
  
  const [headerB64, payloadB64, signatureB64] = parts;
  const signingInput = `${headerB64}.${payloadB64}`;
  
  const key = await crypto.subtle.importKey(
    "raw",
    str2ab(secret),
    { name: "HMAC", hash: { name: "SHA-256" } },
    false,
    ["verify"]
  );
  
  let verified;
  try {
    verified = await crypto.subtle.verify(
      "HMAC",
      key,
      base64UrlDecode(signatureB64),
      str2ab(signingInput)
    );
  } catch (e) {
    throw new Error("Signature verification failed due to crypt error: " + e.message);
  }
  
  if (!verified) {
    throw new Error("Signature verification failed");
  }
  
  const payloadBytes = base64UrlDecode(payloadB64);
  const payload = JSON.parse(new TextDecoder().decode(payloadBytes));
  
  if (verifyExp && payload.exp) {
    if ((Date.now() / 1000) > payload.exp) {
      throw new Error("Token expired");
    }
  }
  
  return payload;
}
